helix-term/src/commands.rs

```diff
@@ -360,92 +360,86 @@ fn move_line_down(cx: &mut Context) {
     let count = cx.count();
     let (view, doc) = current!(cx.editor);
     let text = doc.text().slice(..);
-    let selection = doc.selection(view.id).transform(|range| {
+
+    let selection = doc.selection(view.id).clone().transform(|range| {
         movement::move_vertically(text, range, Direction::Forward, count, Movement::Move)
     });
     doc.set_selection(view.id, selection);
 }
 
 fn goto_line_end(cx: &mut Context) {
     let (view, doc) = current!(cx.editor);
+    let text = doc.text().slice(..);
 
-    let selection = doc.selection(view.id).transform(|range| {
-        let text = doc.text();
-        let line = text.char_to_line(range.head);
+    let selection = doc.selection(view.id).clone().transform(|range| {
+        let line = range.head_line(text);
+        let line_start = text.line_to_char(line);
 
-        let pos = line_end_char_index(&text.slice(..), line);
-        let pos = graphemes::nth_prev_grapheme_boundary(text.slice(..), pos, 1);
-        let pos = range.head.max(pos).max(text.line_to_char(line));
+        let pos = graphemes::prev_grapheme_boundary(text, line_end_char_index(&text, line))
+            .max(line_start);
 
-        Range::new(
-            match doc.mode {
-                Mode::Normal | Mode::Insert => pos,
-                Mode::Select => range.anchor,
-            },
-            pos,
-        )
+        if doc.mode == Mode::Select {
+            range.move_head(text, pos, true)
+        } else {
+            Range::point(pos)
+        }
```

Comment:
I think this type of pattern should be extracted, it's essentially `if cond == true extend else move`

Reply:
Ah, good call!